internal/controller/namespacelabel_controller_test.go

```diff
@@ -0,0 +1,188 @@
+		log.Fatalf("Failed to set PROTECTED_LABELS environment variable: %v", err)
+	}
+})
+
+func createNamespace(name string) {
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+	Expect(k8sClient.Create(ctx, namespace)).To(Succeed())
+}
+
+func deleteNamespace(name string) {
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+	Expect(k8sClient.Delete(ctx, namespace)).To(Succeed())
+}
+
+func deleteAllNamespaceLabels() {
+	namespaceLabelList := &labelsv1.NamespacelabelList{}
+	Expect(k8sClient.List(ctx, namespaceLabelList)).To(Succeed())
+	for _, nl := range namespaceLabelList.Items {
+		Expect(k8sClient.Delete(ctx, &nl)).To(Succeed())
+	}
+}
+
+func getNextEvent(recorder *record.FakeRecorder) string {
+	select {
+	case event := <-recorder.Events:
+		return event
+	default:
+		return ""
+	}
+}
+
+var _ = Describe("NamespaceLabel Controller", func() {
+	BeforeEach(func() {
+		By("Creating default namespace")
+		createNamespace("default")
+		deleteAllNamespaceLabels()
+	})
+
+	AfterEach(func() {
+		By("Cleaning up test resources")
+		deleteAllNamespaceLabels()
+		deleteNamespace("default")
+	})
+
+	Context("When reconciling multiple NamespaceLabel resources in the same namespace", func() {
+		const namespaceName = "default"
+
+		It("should apply only non-overlapping labels from multiple NamespaceLabel resources", func() {
+			By("creating the first NamespaceLabel resource")
+			firstNamespaceLabel := &labelsv1.Namespacelabel{
+				ObjectMeta: metav1.ObjectMeta{Name: "test-resource-unique-1", Namespace: namespaceName},
+				Spec: labelsv1.NamespacelabelSpec{
+					Labels: map[string]string{"key1": "value1", "key2": "value2"},
+				},
+			}
+			Expect(k8sClient.Create(ctx, firstNamespaceLabel)).To(Succeed())
+
+			By("creating the second NamespaceLabel resource with overlapping labels")
+			secondNamespaceLabel := &labelsv1.Namespacelabel{
+				ObjectMeta: metav1.ObjectMeta{Name: "test-resource-unique-2", Namespace: namespaceName},
+				Spec: labelsv1.NamespacelabelSpec{
+					Labels: map[string]string{"key2": "new-value", "key3": "value3"},
+				},
+			}
+			Expect(k8sClient.Create(ctx, secondNamespaceLabel)).To(Succeed())
+
+			recorder := record.NewFakeRecorder(100)
+			reconciler := &NamespacelabelReconciler{
+				Client:   k8sClient,
+				Scheme:   scheme,
+				Log:      zap.New(zap.UseDevMode(true)),
+				Recorder: recorder,
+			}
+
+			By("reconciling the first NamespaceLabel")
+			_, err := reconciler.Reconcile(ctx, ctrl.Request{NamespacedName: types.NamespacedName{Name: "test-resource-unique-1", Namespace: namespaceName}})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("reconciling the second NamespaceLabel")
+			_, err = reconciler.Reconcile(ctx, ctrl.Request{NamespacedName: types.NamespacedName{Name: "test-resource-unique-2", Namespace: namespaceName}})
+			Expect(err).NotTo(HaveOccurred())
```

Comment:
Why are you doing it like this instead of like in the kubebuilder book? This is not how you should test your controllers

https://book.kubebuilder.io/cronjob-tutorial/writing-tests#testing-your-controllers-behavior